src/App/Fossa/Test.hs

```diff
@@ -0,0 +1,174 @@
+      , pretty $ Fossa.renderIssueType ty
+      , "========================================================================"
+      , hsep $ map (fill padding) $ case ty of
+          Fossa.IssuePolicyConflict -> ["Dependency", "Revision", "License"]
+          Fossa.IssuePolicyFlag -> ["Dependency", "Revision", "License"]
+          _ -> ["Dependency", "Revision"]
+      , ""
+      ]
+
+    renderIssue :: Fossa.Issue -> Doc ann
+    renderIssue issue = hsep (map format [name, revision, license])
+      where
+        format :: Text -> Doc ann
+        format = fill padding . pretty
+
+        locatorSplit = T.split (\c -> c == '$' || c == '+') (Fossa.issueRevisionId issue)
+
+        name = fromMaybe (Fossa.issueRevisionId issue) (locatorSplit !? 1)
+        revision = fromMaybe "" (locatorSplit !? 2)
+        license = fromMaybe "" (Fossa.ruleLicenseId =<< Fossa.issueRule issue)
+
+        (!?) :: [a] -> Int -> Maybe a
+        xs !? ix
+          | length xs <= ix = Nothing
+          | otherwise = Just (xs Unsafe.!! ix)
+
+data TestError
+  = TestErrorAPI Fossa.FossaError -- ^ we encountered an API request error
+  | TestBuildFailed -- ^ we encountered the FAILED status on a build
+  deriving (Show, Generic)
+
+renderTestError :: TestError -> Text
+renderTestError (TestErrorAPI err) = "An API error occurred: " <> T.pack (show err)
+renderTestError TestBuildFailed = "The build failed. Check the FOSSA webapp for more details."
+
+waitForBuild
+  :: (Has (Error TestError) sig m, MonadIO m, Has Logger sig m)
+  => Url 'Https
+  -> Text -- ^ api key
+  -> Fossa.ProjectRevision
+  -> m ()
+waitForBuild baseurl key revision = do
+  maybeBuild <- Fossa.fossaReq $ Fossa.getLatestBuild baseurl key revision
+  case maybeBuild of
+    Left err -> throwError (TestErrorAPI err)
+    Right build -> do
+      case Fossa.buildTaskStatus (Fossa.buildTask build) of
+        Fossa.StatusSucceeded -> pure ()
+        Fossa.StatusFailed -> throwError TestBuildFailed
+        otherStatus -> do
+          logSticky $ "[ Waiting for build completion... last status: " <> viaShow otherStatus <> " ]"
+          liftIO $ threadDelay (pollDelaySeconds * 1_000_000)
+          waitForBuild baseurl key revision
+
+waitForIssues
+  :: (Has (Error TestError) sig m, MonadIO m, Has Logger sig m)
+  => Url 'Https
+  -> Text -- ^ api key
+  -> Fossa.ProjectRevision
+  -> m Fossa.Issues
+waitForIssues baseurl key revision = do
+  result <- Fossa.fossaReq $ Fossa.getIssues baseurl key revision
+  case result of
+    Left err -> throwError (TestErrorAPI err)
+    Right issues ->
+      case Fossa.issuesStatus issues of
+        "WAITING" -> do
+          liftIO $ threadDelay (pollDelaySeconds * 1_000_000)
+          waitForIssues baseurl key revision
+        _ -> pure issues
+
+timeout
+  :: Int -- ^ number of seconds before timeout
+  -> IO a
+  -> IO (Maybe a)
+timeout seconds act = either id id <$> Async.race (Just <$> act) (threadDelay (seconds * 1_000_000) *> pure Nothing)
```

Comment:
`Control.Concurrent.Async` gives us a `race` primitive (run two actions, returning the value from whichever finishes first), which allows us to pretty cleanly write a timeout function.